### PR TITLE
For #25722 - Refactor tabs tray interactors into one interactor

### LINKED ISSUE #25491: Refactor InactiveTabs' interactors
Per https://github.com/mozilla-mobile/fenix/pull/24735#discussion_r886194314 and https://github.com/mozilla-mobile/fenix/pull/24735#discussion_r886199844, the Inactive Tabs controllers need to be refactored:

- The delete all functionality should live within the `InactiveTabsInteractor` call stack and not `TabsTrayInteractor`
- `InactiveTabsController` should instead have a `InactiveTabsController` interface and a `DefaultInactiveTabsController` implementor
- The calls to `BrowserTrayInteractor` should be put into `InactiveTabsInteractor`

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20633)

### LINKED ISSUE #25722: Refactor Tabs Tray's interactors
Per https://github.com/mozilla-mobile/fenix/pull/25636#discussion_r903805714 and https://github.com/mozilla-mobile/fenix/pull/25636#discussion_r903843844, the Tabs Tray's interactors should be consolidated into one top-level interactor so there is a clearer call stack for handling user interactions. This way there is one entry point for everything Tabs Tray-related functionality and only one dependency passed down into front-end code consumers.

This effort should be split into the following issues:

**Phase 1 (minor refactoring noise and quick wins**
- [x] https://github.com/mozilla-mobile/fenix/issues/28261
- [x] https://github.com/mozilla-mobile/fenix/issues/28262
- [x] https://github.com/mozilla-mobile/fenix/issues/28263
- [x] https://github.com/mozilla-mobile/fenix/issues/28264
- [x] https://github.com/mozilla-mobile/fenix/issues/28516

**Phase 2 (substantive refactors)**

- [ ] https://github.com/mozilla-mobile/fenix/issues/28265
- [x] https://github.com/mozilla-mobile/fenix/issues/28266
- [x] https://github.com/mozilla-mobile/fenix/issues/28267
- [x] https://github.com/mozilla-mobile/fenix/issues/28268
- [x] https://github.com/mozilla-mobile/fenix/issues/26017

**Phase 3 (follow-up)
- [ ] Create an interactor/controller for the FAB
- [ ] Refactor the sync tabs portion of the tabs tray FAB to live in the FAB interactor/controller flow
- [ ] Decouple `TabsTrayInteractor` from `TabsTray.Delegate` after #21318 

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20781)

### PR DESCRIPTION
**Pull request by MozillaNoah**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #25722

### CODE REVIEW

**Review by gabrielluong — APPROVED** [review]
My overall feedback is to try to avoid making your PR too big, and consider breaking out the changes to more reviewable chunks if the changes do become too large. For refactoring, I am naturally comparing the old vs new changes to make sure we are still preserving the original logic. In an ideal situation, I should be able to easily compare 2 interactors being collapsed into 1 very simply. It did become a bit more complex because of the renaming and the fact that we were collapsing the browser and navigation interactor altogether in this one.

File: app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayController.kt [hunk 1]
```diff
@@ -130,14 +221,57 @@ class DefaultTabsTrayController(
...
         profiler?.addMarker(
             "DefaultTabTrayController.onNewTabTapped",
             startTime
         )
         sendNewTabEvent(isPrivate)
     }
 
+    override fun handleTabOpen(tabId: String, source: String?) {
+        selectTab.invoke(tabId)
+        TabsTray.openedExistingTab.record(TabsTray.OpenedExistingTabExtra(source ?: "unknown"))
```
[the remark this answers is not in the document]
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayController.kt:234]: Good catch with this and agreed on all points. I was leaning towards having QA validate this before it lands.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/TabGroupListAdapter.kt [hunk 2]
```diff
@@ -34,14 +35,14 @@ import org.mozilla.fenix.tabstray.ext.MIN_COLUMN_WIDTH_DP
...
- * @param interactor [BrowserTrayInteractor] handling tabs interactions in a tab tray.
+ * @param tabsTrayInteractor [TabsTrayInteractor] handling tabs interactions in the tabs tray.
  * @param store [TabsTrayStore] containing the complete state of tabs tray and methods to update that.
  * @param featureName [String] representing the name of the feature displaying tabs. Used in telemetry reporting.
  * @param viewLifecycleOwner [LifecycleOwner] life cycle owner for the view.
  */
 class TabGroupListAdapter(
     private val context: Context,
-    private val interactor: BrowserTrayInteractor,
+    private val tabsTrayInteractor: TabsTrayInteractor,
```
[the remark this answers is not in the document]
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/TabGroupListAdapter.kt]: All of this makes sense. We've already talked offline, but I let this PR runaway with scope. I'll make a more concerted effort at making PRs discrete and self-contained as possible. Thanks for the feedback and suggestions.

File: app/src/main/java/org/mozilla/fenix/tabstray/FloatingActionButtonBinding.kt [hunk 3]
```diff
@@ -21,7 +19,7 @@ import org.mozilla.fenix.tabstray.browser.BrowserTrayInteractor
 class FloatingActionButtonBinding(
     private val store: TabsTrayStore,
     private val actionButton: ExtendedFloatingActionButton,
-    private val browserTrayInteractor: BrowserTrayInteractor
+    private val tabsTrayInteractor: TabsTrayInteractor,
```
[the remark this answers is not in the document]
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/tabstray/FloatingActionButtonBinding.kt]: I think this makes sense. I've updated all of the variables/named parameters.

[the remark this answers is not in the document]
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/tabstray/FloatingActionButtonBinding.kt]: I think this makes sense. I've updated all of the variables/named parameters.

File: app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayController.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayController.kt:234]: https://github.com/mozilla-mobile/fenix/issues/26017

**Comment by JohanLorenzo** [comment]
Hey there! I'm sorry to jump in this PR. I'm just rebasing it because https://github.com/mozilla-mobile/fenix/pull/26073 reached `main`. It renamed the expected check from `pr-complete` to `complete-pr`. This impacts mergify and it may prevent you from merging this PR. That's why the rebase is needed. Let me know if you have any questions.

**Comment by MozillaNoah** [comment]
getting an intermittent failure on `TEST: WHEN record is called THEN telemetry is recorded with the appropriate label`

retrying build

**Comment by MozillaNoah** [comment]
This PR is a little old and became a bit unwieldy, so I'm closing this in favor of opening some newer, smaller PRs which cover the refactor pain points covered here.

**Pull request closed without merging**